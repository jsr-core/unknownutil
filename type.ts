/**
 * A type predicate function.
 *
 * ```ts
 * import { as, is, type Predicate } from "@core/unknownutil";
 *
 * type Person = {
 *   name: string;
 *   age: number;
 *   address?: string;
 * };
 * const isPerson = is.ObjectOf({
 *   name: is.String,
 *   age: is.Number,
 *   address: as.Optional(is.String),
 * }) satisfies Predicate<Person>;
 * ```
 */
export type Predicate<T> = (x: unknown) => x is T;

/**
 * A type predicated by {@linkcode Predicate<T>}.
 *
 * ```ts
 * import { as, is, type PredicateType } from "@core/unknownutil";
 *
 * const isPerson = is.ObjectOf({
 *   name: is.String,
 *   age: is.Number,
 *   address: as.Optional(is.String),
 * });
 *
 * type Person = PredicateType<typeof isPerson>;
 * // type Person = {
 * //   name: string;
 * //   age: number;
 * //   address?: string;
 * // };
 * ```
 */
export type PredicateType<P> = P extends Predicate<infer T> ? T : never;

/**
 * JavaScript primitive types.
 */
export type Primitive =
  | string
  | number
  | bigint
  | boolean
  | null
  | undefined
  | symbol;
