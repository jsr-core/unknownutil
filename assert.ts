import type { Predicate } from "./type.ts";

/**
 * A factory function that generates assertion error messages.
 */
export type AssertMessageFactory = (
  x: unknown,
  pred: Predicate<unknown>,
  name?: string,
) => string;

/**
 * The default factory function used to generate assertion error messages.
 */
export const defaultAssertMessageFactory: AssertMessageFactory = (
  x,
  pred,
  name,
) => {
  const p = pred.name || "anonymous predicate";
  const t = typeof x;
  const v = JSON.stringify(x, null, 2);
  return `Expected ${
    name ?? "a value"
  } that satisfies the predicate ${p}, got ${t}: ${v}`;
};

let assertMessageFactory = defaultAssertMessageFactory;

/**
 * Represents an error that occurs when an assertion fails.
 */
export class AssertError extends Error {
  /**
   * Constructs a new instance.
   * @param message The error message.
   */
  constructor(message?: string) {
    super(message);

    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, AssertError);
    }

    this.name = this.constructor.name;
  }
}

/**
 * Sets the factory function used to generate assertion error messages.
 *
 * ```ts
 * import { is, setAssertMessageFactory } from "@core/unknownutil";
 *
 * setAssertMessageFactory((x, pred) => {
 *   if (pred === is.String) {
 *     return `Expected a string, got ${typeof x}`;
 *   } else if (pred === is.Number) {
 *     return `Expected a number, got ${typeof x}`;
 *   } else if (pred === is.Boolean) {
 *     return `Expected a boolean, got ${typeof x}`;
 *   } else {
 *     return `Expected a value that satisfies the predicate, got ${typeof x}`;
 *   }
 * });
 * ```
 *
 * @param factory The factory function.
 */
export function setAssertMessageFactory(factory: AssertMessageFactory): void {
  assertMessageFactory = factory;
}

/**
 * Asserts that the given value satisfies the provided predicate.
 *
 * It throws {@linkcode AssertError} if the value does not satisfy the predicate.
 *
 * ```ts
 * import { assert, is } from "@core/unknownutil";
 *
 * const a: unknown = "hello";
 * assert(a, is.String);
 * // a is now narrowed to string
 * ```
 *
 * @param x The value to be asserted.
 * @param pred The predicate function to test the value against.
 * @param options Optional configuration for the assertion.
 * @returns The function has a return type of `asserts x is T` to help TypeScript narrow down the type of `x` after the assertion.
 */
export function assert<T>(
  x: unknown,
  pred: Predicate<T>,
  options: { message?: string; name?: string } = {},
): asserts x is T {
  if (!pred(x)) {
    throw new AssertError(
      options.message ?? assertMessageFactory(x, pred, options.name),
    );
  }
}
