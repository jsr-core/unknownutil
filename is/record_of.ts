import { rewriteName } from "../_funcutil.ts";
import type { Predicate } from "../type.ts";
import { isRecord } from "./record.ts";

/**
 * Return a type predicate function that returns `true` if the type of `x` satisfies `Record<K, T>`.
 *
 * To enhance performance, users are advised to cache the return value of this function and mitigate the creation cost.
 *
 * Note that this function only check if the `x` satisfies the `Record<K, T>` type.
 * Use {@linkcode isRecordObjectOf} instead if you want to check if the `x` is an instance of `Object`.
 *
 * ```ts
 * import { is } from "@core/unknownutil";
 *
 * const isMyType = is.RecordOf(is.Number);
 * const a: unknown = {"a": 0, "b": 1};
 * if (isMyType(a)) {
 *   const _: Record<PropertyKey, number> = a;
 * }
 * ```
 *
 * With predicate function for keys:
 *
 * ```ts
 * import { is } from "@core/unknownutil";
 *
 * const isMyType = is.RecordOf(is.Number, is.String);
 * const a: unknown = {"a": 0, "b": 1};
 * if (isMyType(a)) {
 *   const _: Record<string, number> = a;
 * }
 * ```
 */
export function isRecordOf<T, K extends PropertyKey = PropertyKey>(
  pred: Predicate<T>,
  predKey?: Predicate<K>,
): Predicate<Record<K, T>> {
  return rewriteName(
    (x: unknown): x is Record<K, T> => {
      if (!isRecord(x)) return false;
      for (const k in x) {
        if (!pred(x[k])) return false;
        if (predKey && !predKey(k)) return false;
      }
      return true;
    },
    "isRecordOf",
    pred,
    predKey,
  );
}
