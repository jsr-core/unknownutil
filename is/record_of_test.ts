import { assertEquals } from "@std/assert";
import { assertSnapshot } from "@std/testing/snapshot";
import { assertType } from "@std/testing/types";
import type { Equal } from "../_testutil.ts";
import { is } from "./mod.ts";
import { isRecordOf } from "./record_of.ts";

Deno.test("isRecordOf<T>", async (t) => {
  await t.step("returns properly named predicate function", async (t) => {
    await assertSnapshot(t, isRecordOf(is.Number).name);
    await assertSnapshot(t, isRecordOf((_x): _x is string => false).name);
  });

  await t.step("returns true on T record", () => {
    assertEquals(isRecordOf(is.Number)({ a: 0 }), true);
    assertEquals(isRecordOf(is.String)({ a: "a" }), true);
    assertEquals(isRecordOf(is.Boolean)({ a: true }), true);
  });

  await t.step("returns false on non T record", () => {
    assertEquals(isRecordOf(is.String)({ a: 0 }), false);
    assertEquals(isRecordOf(is.Number)({ a: "a" }), false);
    assertEquals(isRecordOf(is.String)({ a: true }), false);
  });

  await t.step("predicated type is correct", () => {
    const a: unknown = { a: 0 };
    if (isRecordOf(is.Number)(a)) {
      assertType<Equal<typeof a, Record<PropertyKey, number>>>(true);
    }
  });

  await t.step("checks only object's own properties", async (t) => {
    await t.step("returns true on T record", () => {
      assertEquals(
        isRecordOf(is.Number)(
          Object.assign(Object.create({ p: "ignore" }), { a: 0 }),
        ),
        true,
      );
      assertEquals(
        isRecordOf(is.String)(
          Object.assign(Object.create({ p: 0 /* ignore */ }), { a: "a" }),
        ),
        true,
      );
      assertEquals(
        isRecordOf(is.Boolean)(
          Object.assign(Object.create({ p: "ignore" }), { a: true }),
        ),
        true,
      );
    });
  });

  await t.step("with symbol properties", async (t) => {
    const a = Symbol("a");
    await t.step("returns true on T record", () => {
      assertEquals(isRecordOf(is.Number)({ [a]: 0 }), true);
      assertEquals(isRecordOf(is.String)({ [a]: "a" }), true);
      assertEquals(isRecordOf(is.Boolean)({ [a]: true }), true);
    });

    await t.step("returns false on non T record", () => {
      assertEquals(isRecordOf(is.String)({ [a]: 0 }), false);
      assertEquals(isRecordOf(is.Number)({ [a]: "a" }), false);
      assertEquals(isRecordOf(is.String)({ [a]: true }), false);
    });
  });
});

Deno.test("isRecordOf<T, K>", async (t) => {
  await t.step("returns properly named predicate function", async (t) => {
    await assertSnapshot(t, isRecordOf(is.Number, is.String).name);
    await assertSnapshot(
      t,
      isRecordOf((_x): _x is string => false, is.String).name,
    );
  });

  await t.step("returns true on T record", () => {
    assertEquals(isRecordOf(is.Number, is.String)({ a: 0 }), true);
    assertEquals(isRecordOf(is.String, is.String)({ a: "a" }), true);
    assertEquals(isRecordOf(is.Boolean, is.String)({ a: true }), true);
  });

  await t.step("returns false on non T record", () => {
    assertEquals(isRecordOf(is.String, is.String)({ a: 0 }), false);
    assertEquals(isRecordOf(is.Number, is.String)({ a: "a" }), false);
    assertEquals(isRecordOf(is.String, is.String)({ a: true }), false);
  });

  await t.step("returns false on non K record", () => {
    assertEquals(isRecordOf(is.Number, is.Number)({ a: 0 }), false);
    assertEquals(isRecordOf(is.String, is.Number)({ a: "a" }), false);
    assertEquals(isRecordOf(is.Boolean, is.Number)({ a: true }), false);
  });

  await t.step("predicated type is correct", () => {
    const a: unknown = { a: 0 };
    if (isRecordOf(is.Number, is.String)(a)) {
      assertType<Equal<typeof a, Record<string, number>>>(true);
    }
  });

  await t.step("checks only object's own properties", async (t) => {
    const s = Symbol("s");
    await t.step("returns true on T record", () => {
      assertEquals(
        isRecordOf(is.Number, is.String)(
          Object.assign(Object.create({ p: "ignore" }), { a: 0 }),
        ),
        true,
      );
      assertEquals(
        isRecordOf(is.String, is.String)(
          Object.assign(Object.create({ p: 0 /* ignore */ }), { a: "a" }),
        ),
        true,
      );
      assertEquals(
        isRecordOf(is.Boolean, is.String)(
          Object.assign(Object.create({ p: "ignore" }), { a: true }),
        ),
        true,
      );
      assertEquals(
        isRecordOf(is.String, is.Number)(
          Object.assign(Object.create({ p: "ignore" }), {/* empty */}),
        ),
        true,
        "No own properties",
      );
      assertEquals(
        isRecordOf(is.String, is.String)(
          Object.assign(Object.create({ [s]: "ignore" }), {/* empty */}),
        ),
        true,
        "No own properties",
      );
    });
  });

  await t.step("with symbol properties", async (t) => {
    const a = Symbol("a");
    await t.step("returns properly named predicate function", async (t) => {
      await assertSnapshot(t, isRecordOf(is.Number, is.Symbol).name);
      await assertSnapshot(
        t,
        isRecordOf((_x): _x is string => false, is.Symbol).name,
      );
    });

    await t.step("returns true on T record", () => {
      assertEquals(isRecordOf(is.Number, is.Symbol)({ [a]: 0 }), true);
      assertEquals(isRecordOf(is.String, is.Symbol)({ [a]: "a" }), true);
      assertEquals(isRecordOf(is.Boolean, is.Symbol)({ [a]: true }), true);
    });

    await t.step("returns false on non T record", () => {
      assertEquals(isRecordOf(is.String, is.Symbol)({ [a]: 0 }), false);
      assertEquals(isRecordOf(is.Number, is.Symbol)({ [a]: "a" }), false);
      assertEquals(isRecordOf(is.String, is.Symbol)({ [a]: true }), false);
    });

    await t.step("returns false on non K record", () => {
      assertEquals(isRecordOf(is.Number, is.String)({ [a]: 0 }), false);
      assertEquals(isRecordOf(is.String, is.String)({ [a]: "a" }), false);
      assertEquals(isRecordOf(is.Boolean, is.String)({ [a]: true }), false);
    });

    await t.step("predicated type is correct", () => {
      const a: unknown = { a: 0 };
      if (isRecordOf(is.Number, is.Symbol)(a)) {
        assertType<Equal<typeof a, Record<symbol, number>>>(true);
      }
    });
  });
});
