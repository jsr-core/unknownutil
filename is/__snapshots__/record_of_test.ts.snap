export const snapshot = {};

snapshot[`isRecordOf<T> > returns properly named predicate function 1`] = `"isRecordOf(isNumber, undefined)"`;

snapshot[`isRecordOf<T> > returns properly named predicate function 2`] = `"isRecordOf((anonymous), undefined)"`;

snapshot[`isRecordOf<T, K> > returns properly named predicate function 1`] = `"isRecordOf(isNumber, isString)"`;

snapshot[`isRecordOf<T, K> > returns properly named predicate function 2`] = `"isRecordOf((anonymous), isString)"`;

snapshot[`isRecordOf<T, K> > with symbol properties > returns properly named predicate function 1`] = `"isRecordOf(isNumber, isSymbol)"`;

snapshot[`isRecordOf<T, K> > with symbol properties > returns properly named predicate function 2`] = `"isRecordOf((anonymous), isSymbol)"`;
